utils/camera/camera_driver.cc

```diff
@@ -0,0 +1,80 @@
+#include "camera_driver.h"
+#include <thread>
+#include <chrono>
+#include <iostream>
+#include <string>
+
+namespace camera {
+
+CameraBase::CameraBase(uint32_t n) {
+    this->n = n;
+    this->busy_index = 0;
+    this->is_capturing = false;
+    this->_buffer.resize(this->n);
+}
+
+void CameraBase::_start_capture(unsigned int sleep) {
+    if (this->is_capturing) {
+        // TODO: add legit exception handling
+        std::cerr << "DONT CALL ME TWICE!!!!!\r\n";
+    }
+    this->is_capturing = true;
+    while (true) {
+        this->_buffer[this->busy_index] = this->cam_read();
+        this->_index_lock.lock();
+        this->busy_index = (this->busy_index + 1) % this->n;
+        this->_index_lock.unlock();
+        if (sleep)
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleep));
+    }
+}
+
+void CameraBase::get_img(cv::Mat &dst) {
+    this->_index_lock.lock();
+    uint32_t ind_to_read = (this->busy_index + this->n - 1) % this->n;
+    // this can be potentially improved by taking (diminishing) tradeoff with thread safety
+    // but on Jetson TX2, this implementation already runs faster than 120fps
+    // which is faster than almost every camera...
+    this->_buffer[ind_to_read].copyTo(dst);
```

Comment:
If the circular buffer is big enough and the processing frequency is faster than the circular capture frequency, we could get rid of this copy.